keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2346 @@
+have resources available.
+
+To gather information about the current state of resource availability, the
+scheduler creates a PodScheduling object. That object is owned by the pod and
+will either get deleted by the scheduler when it is done with pod scheduling or
+through the garbage collector. In the PodScheduling object, the scheduler posts
+the list of all potential nodes that it was left with after considering all
+other pod constraints and requirements. Resource drivers involved in the
+scheduling of the pod respond by adding which of these nodes currently don't
+have sufficient resources available. The next scheduling attempt is then more
+likely to pick a node for which allocation succeeds.
+
+This scheduling information is optional and does not have to be in sync with
+the current ResourceClaim state, therefore it is okay to store it
+separately. Alternatively, the scheduler may be configured to do node filtering
+through scheduler extenders, in which case the PodScheduling object will not be
+needed.
+
+Allowing the scheduler to trigger allocation in parallel to asking for more
+information was chosen because for pods with a single resource claim, the cost
+of guessing wrong is low: the driver just needs to inform the scheduler to try
+again and provide the additional information.
+
+Additional heuristics are possible without changing the proposed API. For
+example, the scheduler might ask for information and wait a while before
+making a choice. This may be more suitable for pods using many different
+resource claims because for those, allocation may succeed for some claims and
+fail for others, which then may need to go through the recovery flow with
+deallocating one or more claims.
+
+### Resource allocation and usage flow
+
+The following steps shows how resource allocation works for a resource that
+gets defined inline in a Pod. Several of these steps may fail without changing
+the system state. They then must be retried until they succeed or something
+else changes in the system, like for example deleting objects.
+
+* **user** creates Pod with inline ResourceClaimTemplate
+* **resource claim controller** checks ResourceClaimTemplate and ResourceClass,
+  then creates ResourceClaim with Pod as owner
+* if *immediate allocation*:
+  * **resource driver** adds finalizer to claim to prevent deletion -> allocation in progress
+  * **resource driver** finishes allocation, sets `Allocation` -> claim ready for use by any pod
+* if *pod is pending*:
+  * **scheduler** filters nodes
+  * if *delayed allocation and resource not allocated yet*:
+    * if *at least one node fits pod*:
+      * **scheduler** creates or updates a `PodScheduling` object with `PotentialNodes=<all nodes that fit the pod>`
+      * **scheduler** picks one node, sets `ResourceClaimStatus.Scheduling.SelectedNode=<the chosen node>` and `ResourceClaimStatus.Scheduling.SelectedUser=<the pod being scheduled>`
```

Comment:
I had kept `SelectedNode` in the ResourceClaimStatus because I was worried about race conditions. In the current revision, ResourceClaimStatus holds all fields which must be updated atomically and PodScheduling the optional ones for node selection (but not the actual allocation or deallocation!).

Aldo also pointed this out, so I already started revising the KEP with `SelectedNode` in PodScheduling. But I am indeed running into problems with that because of races and I am not sure yet whether those can be resolved, so I need to continue with that tomorrow.

Reply:
I will have to stare at this even harder.  I think I get what you are doing, but I feel like there has to be a way to do all of the coordination in the coordination object without triggering any actual allocations until we have a solution.